### PR TITLE
Allow templates to be passed as symbols to macros

### PR DESCRIPTION
**Pull request by zah**
Please see the commit messages for a more detailed explanation and rationale.

### CODE REVIEW

**Comment by Araq** [comment]
The compiler additions look OK, but I don't understand your logging module example. What's the point?

**Comment by zah** [comment]
The goal of the logging module is to allow you to define a set of properties that will be automatically associated with each log statement in the lexical scope of your definition. This assumes that logger is able to log such properties as data (you've seen our team's private discussion regarding this).

You can think of this as a solution to a more general problem - how to bind a compile-time value to a particular scope and how to have that value later amendable in sub-scopes or even the same scope. The automatically inserted logger properties are just a specific example of this problem.

**Comment by Araq** [comment]
Fails with:

 lib/pure/terminal.nim(495, 8) Error: undeclared identifier: 'styleCache'

**Comment by zah** [comment]
I Rebased again and used the opportunity to fix the `terminal` module.

**Comment by Araq** [comment]
CIs are still not happy:

```
lib\pure\terminal.nim(496, 19) template/generic instantiation from here
lib\pure\strformat.nim(313, 39) Error: undeclared identifier: 'stylePrefix'
FAILURE
```

File: lib/pure/terminal.nim [hunk 1]
```diff
@@ -29,9 +29,7 @@ when not hasThreadSupport:
   var
     colorsFGCache = initTable[Color, string]()
     colorsBGCache = initTable[Color, string]()
-  when not defined(windows):
```
Comment (Araq) [inline lib/pure/terminal.nim]: There was a reason this was disabled on Windows.

Comment (zah) [reply lib/pure/terminal.nim]: Well, the reason is not valid anymore. ANSI codes are now supported by some Windows terminals and I've exported the procs for generating them for Windows targets too.

Comment (data-man) [reply lib/pure/terminal.nim]: It's disabled because the styleCache table for Windows is not used and not needed.

File: lib/pure/terminal.nim [hunk 2]
```diff
@@ -621,15 +636,19 @@ proc getBGColorStr(color: Color): string =
       result = fmt"{bgPrefix}{rgb.r};{rgb.g};{rgb.b}m"
       colorsFGCache[color] = result
 
+template ansiBackgroundColorCode*(color: static[Color]): string =
+  const rgb = extractRGB(color)
```
Comment (data-man) [inline lib/pure/terminal.nim]: In ansiForegroundColorCode template same code as ``` let rgb = extractRGB(color)```

Comment (zah) [reply lib/pure/terminal.nim]: I'm not sure I understand the question. The goal of this overload is to intercept the case when the color is known at compile-time. In this case, it's possible to compute the final ANSI code right-away at compile-time. 

Without the overload, the run-time cache in the regular overload will prevent the compile-time evaluation.

Comment (zah) [reply lib/pure/terminal.nim]: You may be asking why I used `const` in one situation and `let` in another similar situation. You are right, I should have used `const` in both places.

**Pull request merged**